[PyTorch] Devirtualize TensorImpl::storage()

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#51050 [PyTorch] Devirtualize TensorImpl::storage()**

Subclasses want to be able to make storage() calls throw, so
we find some free space in TensorImpl to add a flag that they can set
to make that happen without making storage() virtual. It should still
be inlineable.

Differential Revision: [D26013815](https://our.internmc.facebook.com/intern/diff/D26013815/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D26013815/)!

swolchok committed:
[PyTorch] Devirtualize TensorImpl::storage()

Subclasses want to be able to make storage() calls throw, so
we find some free space in TensorImpl to add a flag that they can set
to make that happen without making storage() virtual. It should still
be inlineable.

Differential Revision: [D26013815](https://our.internmc.facebook.com/intern/diff/D26013815/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D26013815/)!

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Devirtualize TensorImpl::storage()"


Subclasses want to be able to make storage() calls throw, so
we find some free space in TensorImpl to add a flag that they can set
to make that happen without making storage() virtual. It should still
be inlineable.

Differential Revision: [D26013815](https://our.internmc.facebook.com/intern/diff/D26013815/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D26013815/)!

[ghstack-poisoned]

ezyang commented on c10/core/TensorImpl.h:
```diff
@@ -452,7 +452,12 @@ struct C10_API TensorImpl : public c10::intrusive_ptr_target {
+  TENSORIMPL_MAYBE_VIRTUAL const Storage& storage() const {
+    if (C10_UNLIKELY(storage_access_should_throw_)) {
+      throw_storage_access_error();
+    }
```
If I were king we'd just return a nullptr in this case :> :> :>

ezyang reviewed (approved):
This is a very safe change.

swolchok committed:
Update on "[PyTorch] Devirtualize TensorImpl::storage()"


Subclasses want to be able to make storage() calls throw, so
we find some free space in TensorImpl to add a flag that they can set
to make that happen without making storage() virtual. It should still
be inlineable.

Differential Revision: [D26013815](https://our.internmc.facebook.com/intern/diff/D26013815/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D26013815/)!

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Devirtualize TensorImpl::storage()"


Subclasses want to be able to make storage() calls throw, so
we find some free space in TensorImpl to add a flag that they can set
to make that happen without making storage() virtual. It should still
be inlineable.

Differential Revision: [D26013815](https://our.internmc.facebook.com/intern/diff/D26013815/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D26013815/)!

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Devirtualize TensorImpl::storage()"


Subclasses want to be able to make storage() calls throw, so
we find some free space in TensorImpl to add a flag that they can set
to make that happen without making storage() virtual. It should still
be inlineable.

Differential Revision: [D26013815](https://our.internmc.facebook.com/intern/diff/D26013815/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D26013815/)!

[ghstack-poisoned]

swolchok committed:
fix SparseTensorImpl on "[PyTorch] Devirtualize TensorImpl::storage()"


Subclasses want to be able to make storage() calls throw, so
we find some free space in TensorImpl to add a flag that they can set
to make that happen without making storage() virtual. It should still
be inlineable.

Differential Revision: [D26013815](https://our.internmc.facebook.com/intern/diff/D26013815/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D26013815/)!

[ghstack-poisoned]
